solo-io/gloo
# plugin/als: Access logging can be configured in multiple places this enables another level of it

# Description

https://www.envoyproxy.io/docs/envoy/latest/configuration/observability/access_log/usage
Command operators have special usage in different configurations. This includes multiple places for an http listener. This change adds the ability to add the listener level access logging to http listeners rather than just the HTTP connection manager level.

_Fill out any of the following sections that are relevant and remove the others_
## API changes
- Added EarlyAccessLoggingService to listener options which mimicks the existing AccessLoggingService option in the same location

## Code changes
- Added checks for non-usable command operators.
- als plugin is now also a ListenerPlugin and can update listeners



# Context

Users want to debug some early failures that should be caught with listener access logs.

## Testing steps

I ran my local tests. 
I broke my local tests (including non-ginkgo) and validated it failed tests
Built an image and installed locally to make sure that we had access logs and could use the new directive.

## Notes for reviewers

Be sure to validate that DOWNSTREAM_TRANSPORT_FAILURE_REASON works with listener level logging. This was done manually but given churn of operators we are not currently proposing it for an automatic test.

Please add opinions for naming and helm configuration. 
Today we support high level configuration that will pass access logs to default gateways. In this instance I've opted to not do that due to this being more of an edge case but happy to change that if preferred.

# Checklist:
